File: src/1320-Minimum-Distance-to-Type-a-Word-Using-Two-Fingers/1320.py
from functools import lru_cache
class Solution:
    def minimumDistance(self, word: str) -> int:
        n, sum_dis, max_a = len(word), 0, 0
        
        @lru_cache(None)
        def dis(a, b):
            return abs(a//6 - b//6) + abs(a%6 - b%6)
        
        dp = [0] * 26
        for i in range(n - 1):
            a, c = ord(word[i]) - 65, ord(word[i + 1]) - 65
            dp[a] = max(dp[b] + dis(a, c) - dis(b, c) for b in range(26))
            max_a = max(max_a, dp[a])
            sum_dis += dis(a, c)
        return sum_dis - max_a

File: src/1320-Minimum-Distance-to-Type-a-Word-Using-Two-Fingers/1320.cpp
class Solution 
{
public:
    int minimumDistance(string word) 
    {
        vector<int> dp(26);
        int sum_dis = 0, max_a = 0;
        
        for (int i = 0; i < word.size() - 1; i++)
        {
            int a = word[i] - 65, c = word[i + 1] - 65;
            for (int b = 0; b < 26; b++) 
                dp[a] = max(dp[a], dp[b] + dis(a, c) - dis(b, c));
            max_a = max(max_a, dp[a]);
            sum_dis += dis(a, c);
        }
        
        return sum_dis - max_a;
    }
private:
    int dis(int a, int b)
    {
        return abs(a/6 - b/6) + abs(a%6 - b%6);
    }
};  